### ISSUE TITLE
Slowness with Neural Network Example

### ISSUE THREAD

**Issue by thomasnield** [mipt-npm/kmath#63]
Slowness with neural network usage example

I started to port over some existing machine learning from scratch demos to use kmath. This particular neural network demo I made using stochastic hill climbing is extremely slow, much slower than just using vanilla Kotlin.

I'm open to feedback and learning if I'm doing something wrong...

https://github.com/thomasnield/numky/blob/master/src/test/kotlin/org/nield/numky/NeuralNetworkSimulatedAnnealing.kt

**Comment by thomasnield** [comment]
I do sometimes wonder if there's a use case for a mutable matrix implementation, so variables can be mutated rather than rebuilding an entire matrix with value changes. But then maybe a VirtualMatrix should be used to abstract mutations?

**Comment by altavir** [comment]
I will look into it today with profiler. 

As for mutating matrices, there was a long debate about it. After all, I've decided, that it brings more problems than it solves. So matrices (meaning `Structure2D` are read-only by default). Yet for performance reasons, we usually want to have locally mutable 2d structures. For that we have [BufferAccessor2D](https://github.com/mipt-npm/kmath/blob/master/kmath-core/src/commonMain/kotlin/scientifik/kmath/structures/BufferAccessor2D.kt), it is a very low-level instrument which allows operating on a 1d array like it is 2d. You can see some examples [here](https://github.com/mipt-npm/kmath/blob/f706122266e579ffd93957b83f6995992c34a29e/kmath-core/src/commonMain/kotlin/scientifik/kmath/linear/LUPDecomposition.kt#L90). 

Also, if you need a really good performance, you should use one of the libraries for which we have connectors. The API will be the same, you will just have to substitute context (say, use `CMMatrixContext` instead of `RealMatrixContext`).

**Comment by altavir** [comment]
Profiler shows that 98% CPU time is spent in your `softmax` method, so it is not a kmath own issue. I will try to modify it to works faster.

**Comment by altavir** [comment]
The problem is indeed with the `softmax` method. It is not related to kmath, because the bottleneck is sum of exponents of column values. It adds  additional order of complexity to the algorithm. I am not sure how it could be avoided in plain kotlin code (If you have the code, I can try to compare). The only way, to make it faster is to somehow rewrite the algorithm so it would have less calls to this method.

**Comment by thomasnield** [comment]
Dang... okay. I'll need to see how softmax is done in a more numerically efficient manner. 

And I'll go ahead and apply your suggestions. Thanks for looking into this. Benchmarking and performance analysis is definitely not my strong suit.

**Comment by thomasnield** [comment]
You're right, using sigmoid on outer layer causes it to fly much faster. I merge your PR and am playing with it a bit more.

**Comment by altavir** [comment]
Feel free to ping me when you need another profiling.

**Comment by thomasnield** [comment]
@altavir will do and thank you. Fun stuff.
